### ISSUE TITLE
Variable batch-size during on-policy training

### ISSUE THREAD

**Issue by youkaichao** [thu-ml/tianshou#185]
- [ ] I have marked all applicable categories:
    + [ ] exception-raising bug
    + [ ] RL algorithm bug
    + [ ] documentation request (i.e. "X is missing from the documentation.")
    + [x] new feature request
- [x] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [x] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [x] I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, torch, sys
  print(tianshou.__version__, torch.__version__, sys.version)
  0.2.5, 1.5.0, 3.6.10
  ```

On-policy training now splits the replaybuffer into chunks. For 10 transitions with batch-size 4, tianshou yields three batches with batch-size 4, 4, 2, causing the variable batch-size problem.

Generally it is ok. But sometimes it causes problems, especially if we are using BatchNorm during training, which raises an error if the batch-size is 1.

I think there are three possible solutions:

- Discard the one remaining batch. For the example above, just yield two batches with batch-size 4, 4

- Pad the last batch. For the example above, yield three batches with batch-size 4, 4, 4

- Sample batches with the same batch-size. For the example above, sample three batches with batch-size 4 from 10 transitions.

**Comment by Trinkle23897** [comment]
Vote for the last

**Comment by youkaichao** [comment]
Well, in on-policy training, ``process_fn`` also has this problem. Since ``process_fn`` has to compute the return for each state, the last solution is not proper.

The second solution may be better: pad the last batch for forward, and then slice back.

```python
            for b in batch.split(self._batch, shuffle=False):
                original_len = len(b)
                if original_len != self._batch:
                    # pad the final batch
                    b = batch[-self._batch:]
                v_.append(self.critic(b.obs_next)[-original_len:])
                v.append(self.critic(b.obs)[-original_len:])
                old_log_prob.append(self(b).dist.log_prob(
                    to_torch_as(b.act, v[0]))[-original_len:])
```

**Comment by Trinkle23897** [comment]
There is another solution: merge the last batch to the previous one if it is not full. For example, 10 will result in [4, 6] with bsz=4.

**Comment by youkaichao** [comment]
> merge the last batch to the previous one if it is not full. For example, 10 will result in [4, 6] with bsz=4.

Then the behavior of ``split`` should change.

**Comment by Trinkle23897** [comment]
Yes and it won’t affect other parts of the code

**Comment by youkaichao** [comment]
> Then the behavior of split should change.

The current behavior is intuitive and standard. Changing it should be avoided. There is another solution without changing the behavior of ``split``:

```python
remain = None
for i, b in enumerate(list(batch.split(size))[::-1]):
    if i == 0 and len(b) < size:
        remain = b
        continue
    if i == 1:
        b = cat(remain, b)
    # do something with b
# do something if remain is not None and i == 1
```

**Comment by rocknamx8** [comment]
> There is another solution: merge the last batch to the previous one if it is not full. For example, 10 will result in [4, 6] with bsz=4.

I also encountered this problem when using BN.
This solution and the second solution may be better in early training stage in some env (i.e `len(episode) < bsz`).
